File: app/src/main/java/closer/vlllage/com/closer/handler/map/MapSlideFragment.java
package closer.vlllage.com.closer.handler.map;

import android.Manifest;
import android.content.Intent;
import android.location.Address;
import android.location.Geocoder;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.maps.MapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.firebase.iid.FirebaseInstanceId;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import closer.vlllage.com.closer.MapsActivity;
import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.api.models.PhoneResult;
import closer.vlllage.com.closer.api.models.SuggestionResult;
import closer.vlllage.com.closer.handler.bubble.BubbleHandler;
import closer.vlllage.com.closer.handler.bubble.BubbleType;
import closer.vlllage.com.closer.handler.bubble.MapBubble;
import closer.vlllage.com.closer.handler.bubble.MapBubbleMenuItem;
import closer.vlllage.com.closer.handler.bubble.MapBubbleMenuView;
import closer.vlllage.com.closer.handler.data.AccountHandler;
import closer.vlllage.com.closer.handler.data.ApiHandler;
import closer.vlllage.com.closer.handler.data.LocationHandler;
import closer.vlllage.com.closer.handler.data.PermissionHandler;
import closer.vlllage.com.closer.handler.data.PersistenceHandler;
import closer.vlllage.com.closer.handler.data.RefreshHandler;
import closer.vlllage.com.closer.handler.data.SyncHandler;
import closer.vlllage.com.closer.handler.event.EventBubbleHandler;
import closer.vlllage.com.closer.handler.event.EventHandler;
import closer.vlllage.com.closer.handler.group.GroupActivityTransitionHandler;
import closer.vlllage.com.closer.handler.group.GroupMessageAttachmentHandler;
import closer.vlllage.com.closer.handler.group.PhysicalGroupBubbleHandler;
import closer.vlllage.com.closer.handler.group.PhysicalGroupHandler;
import closer.vlllage.com.closer.handler.helpers.ConnectionErrorHandler;
import closer.vlllage.com.closer.handler.helpers.DefaultAlerts;
import closer.vlllage.com.closer.handler.helpers.DisposableHandler;
import closer.vlllage.com.closer.handler.helpers.KeyboardVisibilityHandler;
import closer.vlllage.com.closer.handler.helpers.LatLngStr;
import closer.vlllage.com.closer.handler.helpers.TimerHandler;
import closer.vlllage.com.closer.handler.helpers.Val;
import closer.vlllage.com.closer.handler.helpers.ViewAttributeHandler;
import closer.vlllage.com.closer.pool.PoolFragment;
import closer.vlllage.com.closer.store.models.Event;
import closer.vlllage.com.closer.store.models.Group;
import closer.vlllage.com.closer.store.models.Suggestion;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class MapSlideFragment extends PoolFragment {

    private boolean locationPermissionWasDenied;
    private Runnable pendingRunnable;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.activity_maps, container, false);

        $(NetworkConnectionViewHandler.class).attach(view.findViewById(R.id.connectionError));
        $(ApiHandler.class).setAuthorization($(AccountHandler.class).getPhone());
        $(TimerHandler.class).postDisposable(() -> $(SyncHandler.class).syncAll(), 1325);
        $(TimerHandler.class).postDisposable(() -> $(RefreshHandler.class).refreshAll(), 1625);

        $(BubbleHandler.class).attach(view.findViewById(R.id.bubbleMapLayer), mapBubble -> {
            if ($(MyBubbleHandler.class).isMyBubble(mapBubble)) {
                $(MapActivityHandler.class).goToScreen(MapsActivity.EXTRA_SCREEN_PERSONAL);
            } else {
                $(ReplyLayoutHandler.class).replyTo(mapBubble);
            }
        }, (mapBubble, position) -> {
            $(BubbleHandler.class).remove(mapBubble);
            if (mapBubble.getOnItemClickListener() != null) {
                mapBubble.getOnItemClickListener().onItemClick(position);
            }
        }, mapBubble -> {
            String groupId = ((Event) mapBubble.getTag()).getGroupId();

            if (groupId != null) {
                $(GroupActivityTransitionHandler.class).showGroupMessages(mapBubble.getView(), groupId);
            } else {
                $(DefaultAlerts.class).thatDidntWork();
            }
        }, mapBubble -> {
            $(SuggestionHandler.class).clearSuggestions();
            $(ShareHandler.class).shareTo(mapBubble.getLatLng(), group -> {
                boolean success = false;
                if (mapBubble.getTag() instanceof Suggestion) {
                    Suggestion suggestion = (Suggestion) mapBubble.getTag();
                    success = $(GroupMessageAttachmentHandler.class).shareSuggestion(suggestion, group);
                }

                if (!success) {
                    $(DefaultAlerts.class).thatDidntWork();
                    return;
                }

                $(GroupActivityTransitionHandler.class).showGroupMessages(mapBubble.getView(), group.getId());
            });
        }, mapBubble -> {
            String groupId = ((Group) mapBubble.getTag()).getId();
            $(GroupActivityTransitionHandler.class).showGroupMessages(mapBubble.getView(), groupId);
        });

        $(MapHandler.class).setOnMapReadyListener(map -> $(BubbleHandler.class).attach(map));
        $(MapHandler.class).setOnMapChangedListener(() -> {
            $(BubbleHandler.class).update();
            $(MapZoomHandler.class).update($(MapHandler.class).getZoom());
        });
        $(MapHandler.class).setOnMapClickedListener(latLng -> {
            if ($(ReplyLayoutHandler.class).isVisible()) {
                $(ReplyLayoutHandler.class).showReplyLayout(false);
            } else {
                boolean anyActionTaken;
                anyActionTaken = $(SuggestionHandler.class).clearSuggestions();
                anyActionTaken = anyActionTaken || $(BubbleHandler.class).remove(mapBubble -> BubbleType.MENU.equals(mapBubble.getType()));

                if (!anyActionTaken) {
                    MapBubble menuBubble = new MapBubble(latLng, BubbleType.MENU);
                    menuBubble.setOnItemClickListener(position -> {
                        switch (position) {
                            case 0:
                                $(PhysicalGroupHandler.class).createPhysicalGroup(menuBubble.getLatLng());
                                break;
                            case 1:
                                $(ShareHandler.class).shareTo(menuBubble.getLatLng(), group -> {
                                    boolean success = $(GroupMessageAttachmentHandler.class).shareLocation(menuBubble.getLatLng(), group);

                                    if (!success) {
                                        $(DefaultAlerts.class).thatDidntWork();
                                    } else {
                                        $(GroupActivityTransitionHandler.class).showGroupMessages(menuBubble.getView(), group.getId());
                                    }
                                });
                                break;
                            case 2:
                                $(EventHandler.class).createNewEvent(menuBubble.getLatLng());
                                break;
                        }
                    });
                    $(BubbleHandler.class).add(menuBubble);
                    menuBubble.setOnViewReadyListener(menuBubbleView -> {
                        $(MapBubbleMenuView.class)
                                .getMenuAdapter(menuBubble)
                                .setMenuItems(
                                        new MapBubbleMenuItem().setTitle(getString(R.string.talk_here)).setIconRes(R.drawable.ic_wifi_black_18dp),
                                        new MapBubbleMenuItem().setTitle(getString(R.string.share_this_location)).setIconRes(R.drawable.ic_share_black_18dp),
                                        new MapBubbleMenuItem().setTitle(getString(R.string.add_event_here)).setIconRes(R.drawable.ic_event_note_black_24dp));
                    });
                }
            }
        });
        $(MapHandler.class).setOnMapLongClickedListener(latLng -> {
            $(BubbleHandler.class).remove(mapBubble -> BubbleType.MENU.equals(mapBubble.getType()));

            MapBubble menuBubble = new MapBubble(latLng, BubbleType.MENU);
            menuBubble.setOnItemClickListener(position -> {
                switch (position) {
                    case 0:
                        $(SuggestionHandler.class).createNewSuggestion(menuBubble.getLatLng());
                        break;
                }
            });
            $(BubbleHandler.class).add(menuBubble);
            menuBubble.setOnViewReadyListener(menuBubbleView -> {
                $(MapBubbleMenuView.class)
                        .getMenuAdapter(menuBubble)
                        .setMenuItems(new MapBubbleMenuItem().setTitle(getString(R.string.add_suggestion_here)));
            });
        });
        $(MapHandler.class).setOnMapIdleListener(latLng -> {
            $(DisposableHandler.class).add($(ApiHandler.class).getPhonesNear(latLng).map(this::mapBubbleFrom).subscribe(mapBubbles ->
                    $(BubbleHandler.class).replace(mapBubbles), this::networkError));

            $(DisposableHandler.class).add($(ApiHandler.class).getSuggestionsNear(latLng).map(SuggestionResult::from).subscribe(suggestions ->
                    $(SuggestionHandler.class).loadAll(suggestions), this::networkError));

            $(RefreshHandler.class).refreshEvents(latLng);
            $(RefreshHandler.class).refreshPhysicalGroups(latLng);
        });
        $(MapHandler.class).attach((MapFragment) getChildFragmentManager().findFragmentById(R.id.map));
        $(MyBubbleHandler.class).start();
        $(ReplyLayoutHandler.class).attach(view.findViewById(R.id.replyLayout));
        $(MyGroupsLayoutHandler.class).attach(view.findViewById(R.id.myGroupsLayout));
        $(MyGroupsLayoutHandler.class).setContainerView(view.findViewById(R.id.bottomContainer));

        $(ViewAttributeHandler.class).linkPadding(view.findViewById(R.id.contentAboveView), view.findViewById(R.id.contentView));

        $(KeyboardVisibilityHandler.class).attach(view.findViewById(R.id.contentView));
        $(DisposableHandler.class).add($(KeyboardVisibilityHandler.class).isKeyboardVisible()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(isVisible -> $(MyGroupsLayoutHandler.class).showBottomPadding(!isVisible), Throwable::printStackTrace));

        boolean verifiedNumber = $(PersistenceHandler.class).getIsVerified();
        $(MyGroupsLayoutActionsHandler.class).showVerifyMyNumber(!verifiedNumber);
        if (!verifiedNumber) {
            $(DisposableHandler.class).add($(ApiHandler.class).isVerified().subscribe(verified -> {
                $(PersistenceHandler.class).setIsVerified(verified);
                $(MyGroupsLayoutActionsHandler.class).showVerifyMyNumber(!verified);
            }, this::networkError));
        }

        String deviceToken = FirebaseInstanceId.getInstance().getToken();

        if (deviceToken != null) {
            $(AccountHandler.class).updateDeviceToken(deviceToken);
        }

        if ($(AccountHandler.class).getActive()) {
            $(AccountHandler.class).updateStatus($(AccountHandler.class).getStatus());
        }

        $(MyGroupsLayoutActionsHandler.class).showHelpButton(!$(PersistenceHandler.class).getIsHelpHidden());

        $(EventBubbleHandler.class).attach();
        $(PhysicalGroupBubbleHandler.class).attach();
        $(FeedHandler.class).attach(view.findViewById(R.id.feed));

        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        if (pendingRunnable != null) {
            pendingRunnable.run();
            pendingRunnable = null;
        }
    }

    @Override
    public void onResume() {
        super.onResume();

        boolean locationPermissionGranted = $(PermissionHandler.class).has(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION);
        if (locationPermissionGranted) {
            $(LocationHandler.class).getCurrentLocation(location -> $(AccountHandler.class).updateGeo(new LatLng(location.getLatitude(), location.getLongitude())));
        }

        boolean locationPermissionDenied = $(PermissionHandler.class).denied(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION);
        $(MyGroupsLayoutActionsHandler.class).showAllowLocationPermissionsInSettings(locationPermissionDenied);

        boolean isNotificationsPaused = $(PersistenceHandler.class).getIsNotifcationsPaused();
        $(MyGroupsLayoutActionsHandler.class).showUnmuteNotifications(isNotificationsPaused);
        $(MyGroupsLayoutActionsHandler.class).showSetMyName($(Val.class).isEmpty($(AccountHandler.class).getName()));

        if (locationPermissionGranted && locationPermissionWasDenied) {
            $(MapHandler.class).updateMyLocationEnabled();
            $(MapHandler.class).locateMe();
        }

        locationPermissionWasDenied = locationPermissionDenied;
    }


    public List<MapBubble> mapBubbleFrom(List<PhoneResult> phoneResults) {
        List<MapBubble> mapBubbles = new ArrayList<>();

        for (PhoneResult phoneResult : phoneResults) {
            if (phoneResult.geo == null) {
                continue;
            }

            mapBubbles.add(new MapBubble(
                    $(LatLngStr.class).to(phoneResult.geo),
                    phoneResult.name == null ? "" : phoneResult.name,
                    phoneResult.status
            ).setPhone(phoneResult.id));
        }

        return mapBubbles;
    }

    private void networkError(Throwable throwable) {
        throwable.printStackTrace();
        $(ConnectionErrorHandler.class).notifyConnectionError();
    }

    public boolean onBackPressed() {
        if ($(ReplyLayoutHandler.class).isVisible()) {
            $(ReplyLayoutHandler.class).showReplyLayout(false);
            return true;
        }

        return false;
    }

    public void handleIntent(Intent intent, MapViewHandler.OnRequestMapOnScreenListener onRequestMapOnScreenListener) {
        $(IntentHandler.class).onNewIntent(intent, onRequestMapOnScreenListener);
        if (intent != null) {
            if (Intent.ACTION_VIEW.equals(intent.getAction())) {
                $(FeedHandler.class).hide();
            } else if (Intent.ACTION_SEND.equals(intent.getAction())) {
                String name = intent.getStringExtra(Intent.EXTRA_SUBJECT);
                String address = intent.getStringExtra(Intent.EXTRA_TEXT);

                if (address != null) {
                    $(DisposableHandler.class).add(Single.fromCallable(() -> new Geocoder(getActivity(), Locale.getDefault()).getFromLocationName(address, 1)).subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(addresses -> {
                        if (addresses.isEmpty()) {
                            $(DefaultAlerts.class).thatDidntWork();
                        } else {
                            showAddressOnMap(name, addresses.get(0));
                        }
                    }, this::networkError));

                }
            }
        }
    }

    private void showAddressOnMap(String name, Address address) {
        Suggestion suggestion = new Suggestion();
        suggestion.setName(name);
        suggestion.setLatitude(address.getLatitude());
        suggestion.setLongitude(address.getLongitude());
        $(MapActivityHandler.class).showSuggestionOnMap(suggestion);
    }

    public void post(Runnable runnable) {
        if (isAdded()) {
            runnable.run();
        } else {
            pendingRunnable = runnable;
        }
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/map/MapSlideFragment.kt
package closer.vlllage.com.closer.handler.map

import android.Manifest
import android.content.Intent
import android.location.Address
import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import androidx.core.view.doOnLayout
import androidx.core.view.marginBottom
import androidx.core.view.updateLayoutParams
import androidx.fragment.app.FragmentActivity
import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.extensions.toLatLng
import closer.vlllage.com.closer.handler.bubble.*
import closer.vlllage.com.closer.handler.data.*
import closer.vlllage.com.closer.handler.event.EventBubbleHandler
import closer.vlllage.com.closer.handler.event.EventHandler
import closer.vlllage.com.closer.handler.group.*
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.handler.phone.NameHandler
import closer.vlllage.com.closer.handler.phone.NavigationHandler
import closer.vlllage.com.closer.handler.share.ShareActivityTransitionHandler
import closer.vlllage.com.closer.pool.PoolFragment
import closer.vlllage.com.closer.store.models.Event
import closer.vlllage.com.closer.store.models.Group
import closer.vlllage.com.closer.store.models.Phone
import closer.vlllage.com.closer.store.models.Suggestion
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.firebase.iid.FirebaseInstanceId
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.activity_maps.*
import kotlinx.android.synthetic.main.activity_maps.view.*


class MapSlideFragment : PoolFragment() {

    private var locationPermissionWasDenied: Boolean = false
    private var pendingRunnable: Runnable? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.activity_maps, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val breakpoint = on<ResourcesHandler>().resources.getDimensionPixelSize(R.dimen.maxFullWidth)
        val feedPeekHeightMinus12dp = on<ResourcesHandler>().resources.getDimensionPixelSize(R.dimen.feedPeekHeightMinus12dp)

        mapLayout.addOnLayoutChangeListener { v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom ->
            val w = right - left
            if (mapLayout.marginBottom == 0 && w <= breakpoint) {
                mapLayout.updateLayoutParams<ViewGroup.MarginLayoutParams> {
                    bottomMargin = feedPeekHeightMinus12dp
                }

                mapLayout.requestLayout()
            } else if (mapLayout.marginBottom != 0 && w > breakpoint) {
                mapLayout.updateLayoutParams<ViewGroup.MarginLayoutParams> {
                    bottomMargin = 0
                }

                mapLayout.requestLayout()
            }
        }

        on<NetworkConnectionViewHandler>().attach(view.connectionError)
        on<TimerHandler>().postDisposable({ on<SyncHandler>().syncAll() }, 1325)
        on<TimerHandler>().postDisposable({ on<RefreshHandler>().refreshAll() }, 1625)

        on<BubbleHandler>().attach(view.findViewById(R.id.bubbleMapLayer), { mapBubble ->
            on<NavigationHandler>().showProfile(mapBubble.phone!!, mapBubble.view)
        }, { mapBubble, position ->
                on<BubbleHandler>().remove(mapBubble)
                mapBubble.onItemClickListener?.invoke(position)
        }, { mapBubble ->
            val groupId = (mapBubble.tag as Event).groupId

            if (groupId != null) {
                on<GroupActivityTransitionHandler>().showGroupMessages(mapBubble.view, groupId)
            } else {
                on<DefaultAlerts>().thatDidntWork()
            }
        }, { mapBubble ->
            val suggestion = mapBubble.tag as Suggestion
            if (suggestion.id != null) {
                on<ShareActivityTransitionHandler>().shareSuggestion(suggestion.id!!)
            }
        }, { mapBubble ->
            val groupId = (mapBubble.tag as Group).id
            on<GroupActivityTransitionHandler>().showGroupMessages(mapBubble.view, groupId)
        })

        on<MapHandler>().onMapReadyListener = { map ->
            on<PhysicalGroupBubbleHandler>().attach()
            on<SuggestionBubbleHandler>().attach()
            on<BubbleHandler>().attach(map)
        }
        on<MapHandler>().onMapChangedListener = {
            on<BubbleHandler>().update()
            on<MapZoomHandler>().update(on<MapHandler>().zoom)
        }
        on<MapHandler>().onMapClickedListener = { latLng ->
            val anyActionTaken = on<BubbleHandler>().remove { mapBubble -> BubbleType.MENU == mapBubble.type }

            if (!anyActionTaken) {
                showMapMenu(latLng, null)
            }
        }
        on<MapHandler>().onMapIdleListener = { latLng ->
            on<LocalityHelper>().getLocality(latLng) {
                if (it.isNullOrBlank()) {
                    view.searchMap.hint = on<ResourcesHandler>().resources.getString(R.string.search_map_hint)
                } else {
                    view.searchMap.hint = on<ResourcesHandler>().resources.getString(R.string.search_x_hint, it)
                }
            }

            on<DisposableHandler>().add(on<DataHandler>().run {
                getPhonesNear(latLng)
                        .filter { on<AccountHandler>().privateOnly.not() }
                        .map { mapBubbleFrom(it) }.subscribe({ mapBubbles -> on<BubbleHandler>().replace(mapBubbles) }, { networkError(it) })
            })

            on<DisposableHandler>().add(on<ApiHandler>().getSuggestionsNear(latLng).subscribe({ suggestions -> on<SuggestionHandler>().loadAll(suggestions) }, { networkError(it) }))

            on<RefreshHandler>().refreshEvents(latLng)
            on<RefreshHandler>().refreshPhysicalGroups(latLng)
            on<RefreshHandler>().refreshQuests(latLng)

            on<MapZoomHandler>().update(on<MapHandler>().zoom)

            on<MeetHandler>().setLocation(latLng)
        }
        on<MapHandler>().attach(childFragmentManager.findFragmentById(R.id.map) as SupportMapFragment)
        on<MyBubbleHandler>().start()
        on<MyGroupsLayoutHandler>().attach(view.findViewById(R.id.myGroupsLayout))
        on<MyGroupsLayoutHandler>().setContainerView(view.findViewById(R.id.bottomContainer))

        on<KeyboardVisibilityHandler>().attach(view.findViewById(R.id.contentView))

        on<DisposableHandler>().add(on<KeyboardVisibilityHandler>().isKeyboardVisible
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ isVisible -> on<MyGroupsLayoutHandler>().showBottomPadding(!isVisible) }, { it.printStackTrace() }))

        val verifiedNumber = on<PersistenceHandler>().isVerified
        on<MyGroupsLayoutActionsHandler>().showVerifyMyNumber(!verifiedNumber)
        if (!verifiedNumber) {
            on<DisposableHandler>().add(on<ApiHandler>().isVerified().subscribe({ verified ->
                on<PersistenceHandler>().isVerified = verified
                on<MyGroupsLayoutActionsHandler>().showVerifyMyNumber(!verified)
            }, { networkError(it) }))
        }

        FirebaseInstanceId.getInstance().instanceId.addOnSuccessListener(activity as FragmentActivity) { instanceIdResult ->
            on<AccountHandler>().updateDeviceToken(instanceIdResult.token)
        }

        if (on<AccountHandler>().active) {
            on<AccountHandler>().updateStatus(on<AccountHandler>().status)
        }

        on<EventBubbleHandler>().attach()
        on<FeedHandler>().attach(view.feed, view.toTheTopLayout)

        view.toTheTop.setOnClickListener {
            on<FeedHandler>().hide()
        }

        on<DisposableHandler>().add(on<MeetHandler>().total
                .subscribe {
            on<MyGroupsLayoutActionsHandler>().showMeetPeople(it > 0 && on<AccountHandler>().privateOnly.not())
        })

        if (pendingRunnable != null) {
            pendingRunnable!!.run()
            pendingRunnable = null
        }

        if("I am a super god being".isNotBlank()) {
            on<DataVisualsHandler>().attach()

            on<DisposableHandler>().add(on<DataHandler>().getRecentlyActivePhones(100).map {
                it.filter { it.latitude != null && it.longitude != null }.map { LatLng(it.latitude!!, it.longitude!!) }
            }.subscribe({ phones ->
                on<DataVisualsHandler>().setPhones(phones)
            }, { }))

            on<DisposableHandler>().add(on<DataHandler>().getRecentlyActiveGroups(100).map {
                it.filter { it.latitude != null && it.longitude != null }.map { LatLng(it.latitude!!, it.longitude!!) }
            }.subscribe({ phones ->
                on<DataVisualsHandler>().setGroups(phones)
            }, { }))
        }

        view.searchMap.doOnLayout {
            on<MapHandler>().setTopPadding(view.searchMap.bottom)
        }

        view.locateMeButton.setOnClickListener {
            if (on<PermissionHandler>().denied(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION)) {
                on<AlertHandler>().make().apply {
                    title = on<ResourcesHandler>().resources.getString(R.string.enable_location_permission)
                    message = on<ResourcesHandler>().resources.getString(R.string.enable_location_permission_rationale)
                    positiveButton = on<ResourcesHandler>().resources.getString(R.string.open_settings)
                    positiveButtonCallback = { on<SystemSettingsHandler>().showSystemSettings() }
                    show()
                }

                return@setOnClickListener
            }

            on<LocationHandler>().getCurrentLocation {
                on<MapHandler>().centerMap(it.toLatLng())
            }
        }

        view.scanInviteButton.setOnClickListener {
            on<ScanQrCodeHandler>().scan()
        }

        view.searchMap.setOnEditorActionListener { _, actionId, event ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                search(view.searchMap.text.toString())
                false
            }

            true
        }
    }

    private fun search(query: String) {
        on<AccountHandler>().updatePrivateOnly(false)
        on<SearchMapHandler>().next(query)
    }

    private fun showMapMenu(latLng: LatLng, title: String?) {
        val menuBubble = MapBubble(latLng, BubbleType.MENU, true, true)
        menuBubble.onItemClickListener = { position ->
            when (position) {
                0 -> on<PhysicalGroupHandler>().createPhysicalGroup(menuBubble.latLng!!, name = menuBubble.status ?: "")
                1 -> on<EventHandler>().createNewEvent(menuBubble.latLng!!, true) {
                    on<MapHandler>().centerMap(LatLng(
                            it.latitude!!,
                            it.longitude!!
                    ))
                }
                2 -> on<ShareHandler>().shareTo(menuBubble.latLng!!) { group ->
                    val success = on<GroupMessageAttachmentHandler>().shareLocation(menuBubble.latLng!!, group, menuBubble.status)

                    if (!success) {
                        on<DefaultAlerts>().thatDidntWork()
                    } else {
                        on<GroupActivityTransitionHandler>().showGroupMessages(menuBubble.view, group.id)
                    }
                }
                3 -> on<SuggestionHandler>().createNewSuggestion(menuBubble.latLng!!)
                4 -> {
                    on<DefaultInput>().show(R.string.add_new_public_place, R.string.enter_place_name, R.string.create_place, inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_CAP_WORDS, prefill = menuBubble.status) {
                        on<PhysicalGroupHandler>().createPhysicalGroup(menuBubble.latLng!!, isPublic = true, hub = true, name = it)
                    }
                }
                5 -> {
                    on<DefaultInput>().show(R.string.add_new_private_place, R.string.enter_place_name, R.string.create_place, inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_CAP_WORDS) {
                        on<PhysicalGroupHandler>().createPhysicalGroup(menuBubble.latLng!!, isPublic = false, hub = true, name = it)
                    }
                }
            }
        }
        on<BubbleHandler>().add(menuBubble)
        menuBubble.status = title
        menuBubble.onViewReadyListener = {
            on<MapBubbleMenuView>()
                    .getMenuAdapter(menuBubble)
                    .setMenuItems(
                            (MapBubbleMenuItem(getString(R.string.talk_here), R.drawable.ic_chat_black_18dp, R.color.purple)),
                            (MapBubbleMenuItem(getString(R.string.host_event), R.drawable.ic_event_note_black_18dp, R.color.red)),
                            (MapBubbleMenuItem(getString(R.string.share_location), R.drawable.ic_share_black_18dp, R.color.colorAccent)),
                            (MapBubbleMenuItem(getString(R.string.add_a_suggestion), R.drawable.ic_edit_location_black_18dp, R.color.colorPrimaryLight)),
                            (MapBubbleMenuItem(getString(R.string.add_new_public_place), R.drawable.ic_add_black_18dp, R.color.purple)),
                            (MapBubbleMenuItem(getString(R.string.add_new_private_place), R.drawable.ic_group_add_black_18dp, R.color.colorPrimary)))

            on<MapBubbleMenuView>().setMenuTitle(menuBubble, title ?: on<ResourcesHandler>().resources.getString(R.string.loading_location))

            if (title == null) {
                on<LocalityHelper>().getName(latLng) {
                    menuBubble.status = it ?: on<ResourcesHandler>().resources.getString(R.string.unknown_place)
                    on<MapBubbleMenuView>().setMenuTitle(menuBubble, it ?: on<ResourcesHandler>().resources.getString(R.string.unknown_place))
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()

        val locationPermissionGranted = on<PermissionHandler>().has(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION)
        if (locationPermissionGranted) {
            on<LocationHandler>().getCurrentLocation { location -> on<AccountHandler>().updateGeo(LatLng(location.latitude, location.longitude)) }
        }

        val locationPermissionDenied = on<PermissionHandler>().denied(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION)
        val isNotificationsPaused = on<PersistenceHandler>().isNotificationsPaused
        on<MyGroupsLayoutActionsHandler>().showUnmuteNotifications(isNotificationsPaused)
        on<MyGroupsLayoutActionsHandler>().showSetMyName(on<AccountHandler>().name.isBlank())

        if (locationPermissionGranted && locationPermissionWasDenied) {
            on<MapHandler>().updateMyLocationEnabled()
            on<MapHandler>().locateMe()
        }

        locationPermissionWasDenied = locationPermissionDenied
    }

    private fun mapBubbleFrom(phoneList: List<Phone>): List<MapBubble> {
        val mapBubbles = mutableListOf<MapBubble>()

        for (phone in phoneList) {
            if (phone.latitude == null || phone.longitude == null) {
                continue
            }

            mapBubbles.add(MapBubble(
                    on<LatLngStr>().to(phone.latitude!!, phone.longitude!!),
                    on<NameHandler>().getName(phone),
                    phone.status
            ).apply {
                tag = phone
                this.phone = phone.id
            })
        }

        return mapBubbles
    }

    private fun networkError(throwable: Throwable) {
        throwable.printStackTrace()
        on<ConnectionErrorHandler>().notifyConnectionError()
    }

    fun onBackPressed(): Boolean {
        return false
    }

    fun handleIntent(intent: Intent?, onRequestMapOnScreenListener: (() -> Unit)?) {
        intent ?: return

        on<IntentHandler>().onNewIntent(intent, onRequestMapOnScreenListener)
        if (Intent.ACTION_VIEW == intent.action) {
            on<FeedHandler>().hide()
        } else if (Intent.ACTION_SEND == intent.action) {
            val name = intent.getStringExtra(Intent.EXTRA_SUBJECT)
            val address = intent.getStringExtra(Intent.EXTRA_TEXT)

            if (address != null) {
                on<DisposableHandler>().add(on<PlacesHandler>().findAddress(address, 1)
                        .subscribe({ addresses ->
                            if (addresses.isEmpty()) {
                                on<DefaultAlerts>().thatDidntWork()
                            } else {
                                showAddressOnMap(name ?: on<ResourcesHandler>().resources.getString(R.string.shared_location), addresses[0])
                            }
                        }, { networkError(it) }))

            }
        }
    }

    private fun showAddressOnMap(name: String?, address: Address) {
        val latLng = LatLng(address.latitude, address.longitude)
        showMapMenu(latLng, name)
        on<MapHandler>().centerMap(latLng)
    }

    fun post(runnable: Runnable) {
        if (isAdded) {
            runnable.run()
        } else {
            pendingRunnable = runnable
        }
    }
}
